File: packages/health_connector_hc_android/lib/src/mappers/metadata_mappers.dart
import 'package:health_connector_core/health_connector_core.dart'
    show RecordingMethod, DeviceType, Device, DataOrigin, Metadata, sinceV1_0_0;
import 'package:health_connector_hc_android/src/pigeon/health_connector_platform_api.g.dart'
    show RecordingMethodDto, DeviceTypeDto, MetadataDto;
import 'package:meta/meta.dart' show internal;

/// Converts [RecordingMethod] to [RecordingMethodDto].
@sinceV1_0_0
@internal
extension RecordingMethodDtoMapper on RecordingMethod {
  RecordingMethodDto toDto() {
    switch (this) {
      case RecordingMethod.unknown:
        return RecordingMethodDto.unknown;
      case RecordingMethod.manualEntry:
        return RecordingMethodDto.manualEntry;
      case RecordingMethod.automaticallyRecorded:
        return RecordingMethodDto.automaticallyRecorded;
      case RecordingMethod.activelyRecorded:
        return RecordingMethodDto.activelyRecorded;
    }
  }
}

/// Converts [RecordingMethodDto] to [RecordingMethod].
@sinceV1_0_0
@internal
extension RecordingMethodDtoToDomain on RecordingMethodDto {
  RecordingMethod toDomain() {
    switch (this) {
      case RecordingMethodDto.unknown:
        return RecordingMethod.unknown;
      case RecordingMethodDto.manualEntry:
        return RecordingMethod.manualEntry;
      case RecordingMethodDto.automaticallyRecorded:
        return RecordingMethod.automaticallyRecorded;
      case RecordingMethodDto.activelyRecorded:
        return RecordingMethod.activelyRecorded;
    }
  }
}

/// Converts [DeviceType] to [DeviceTypeDto].
@sinceV1_0_0
@internal
extension DeviceTypeDtoMapper on DeviceType {
  DeviceTypeDto toDto() {
    switch (this) {
      case DeviceType.unknown:
        return DeviceTypeDto.unknown;
      case DeviceType.watch:
        return DeviceTypeDto.watch;
      case DeviceType.phone:
        return DeviceTypeDto.phone;
      case DeviceType.scale:
        return DeviceTypeDto.scale;
      case DeviceType.ring:
        return DeviceTypeDto.ring;
      case DeviceType.fitnessBand:
        return DeviceTypeDto.fitnessBand;
      case DeviceType.chestStrap:
        return DeviceTypeDto.chestStrap;
      case DeviceType.headMounted:
        return DeviceTypeDto.headMounted;
      case DeviceType.smartDisplay:
        return DeviceTypeDto.smartDisplay;
    }
  }
}

/// Converts [DeviceTypeDto] to [DeviceType].
@sinceV1_0_0
@internal
extension DeviceTypeDtoToDomain on DeviceTypeDto {
  DeviceType toDomain() {
    switch (this) {
      case DeviceTypeDto.unknown:
        return DeviceType.unknown;
      case DeviceTypeDto.watch:
        return DeviceType.watch;
      case DeviceTypeDto.phone:
        return DeviceType.phone;
      case DeviceTypeDto.scale:
        return DeviceType.scale;
      case DeviceTypeDto.ring:
        return DeviceType.ring;
      case DeviceTypeDto.fitnessBand:
        return DeviceType.fitnessBand;
      case DeviceTypeDto.chestStrap:
        return DeviceType.chestStrap;
      case DeviceTypeDto.headMounted:
        return DeviceType.headMounted;
      case DeviceTypeDto.smartDisplay:
        return DeviceType.smartDisplay;
    }
  }
}

/// Converts [Metadata] to [MetadataDto].
@sinceV1_0_0
@internal
extension MetadataDtoMapper on Metadata {
  MetadataDto toDto() {
    return MetadataDto(
      dataOrigin: dataOrigin.packageName,
      recordingMethod: recordingMethod.toDto(),
      lastModifiedTime: lastModifiedTime?.millisecondsSinceEpoch,
      clientRecordId: clientRecordId,
      clientRecordVersion: clientRecordVersion,
      deviceType: device?.type.toDto(),
      deviceManufacturer: device?.manufacturer,
      deviceModel: device?.model,
    );
  }
}

/// Converts [MetadataDto] to [Metadata].
@sinceV1_0_0
@internal
extension MetadataDtoToDomain on MetadataDto {
  Metadata toDomain() {
    return Metadata(
      dataOrigin: DataOrigin(dataOrigin),
      recordingMethod: recordingMethod.toDomain(),
      lastModifiedTime: lastModifiedTime != null
          ? DateTime.fromMillisecondsSinceEpoch(lastModifiedTime!)
          : null,
      clientRecordId: clientRecordId,
      clientRecordVersion: clientRecordVersion,
      device:
          (deviceType != null ||
              deviceManufacturer != null ||
              deviceModel != null)
          ? Device(
              type: deviceType?.toDomain() ?? DeviceType.unknown,
              manufacturer: deviceManufacturer,
              model: deviceModel,
            )
          : null,
    );
  }
}

File: packages/health_connector_hc_android/android/src/main/kotlin/com/phamtunglam/health_connector_hc_android/mappers/MetadataMappers.kt
package com.phamtunglam.health_connector_hc_android.mappers

import androidx.health.connect.client.records.metadata.Device
import androidx.health.connect.client.records.metadata.Metadata
import com.phamtunglam.health_connector_hc_android.pigeon.DeviceTypeDto
import com.phamtunglam.health_connector_hc_android.pigeon.MetadataDto
import com.phamtunglam.health_connector_hc_android.pigeon.RecordingMethodDto

/**
 * Converts a Health Connect recording method constant to a [RecordingMethodDto].
 */
internal fun Int.toRecordingMethodDto(): RecordingMethodDto = when (this) {
    Metadata.RECORDING_METHOD_MANUAL_ENTRY -> RecordingMethodDto.MANUAL_ENTRY
    Metadata.RECORDING_METHOD_AUTOMATICALLY_RECORDED -> RecordingMethodDto.AUTOMATICALLY_RECORDED
    Metadata.RECORDING_METHOD_ACTIVELY_RECORDED -> RecordingMethodDto.ACTIVELY_RECORDED
    else -> RecordingMethodDto.UNKNOWN
}

/**
 * Converts a [DeviceTypeDto] to a Health Connect device type constant.
 */
internal fun DeviceTypeDto.toHealthConnect(): Int = when (this) {
    DeviceTypeDto.UNKNOWN -> Device.TYPE_UNKNOWN
    DeviceTypeDto.WATCH -> Device.TYPE_WATCH
    DeviceTypeDto.PHONE -> Device.TYPE_PHONE
    DeviceTypeDto.SCALE -> Device.TYPE_SCALE
    DeviceTypeDto.RING -> Device.TYPE_RING
    DeviceTypeDto.FITNESS_BAND -> Device.TYPE_FITNESS_BAND
    DeviceTypeDto.CHEST_STRAP -> Device.TYPE_CHEST_STRAP
    DeviceTypeDto.HEAD_MOUNTED -> Device.TYPE_HEAD_MOUNTED
    DeviceTypeDto.SMART_DISPLAY -> Device.TYPE_SMART_DISPLAY
}

/**
 * Converts a Health Connect device type constant to a [DeviceTypeDto].
 */
internal fun Int.toDeviceTypeDto(): DeviceTypeDto = when (this) {
    Device.TYPE_UNKNOWN -> DeviceTypeDto.UNKNOWN
    Device.TYPE_WATCH -> DeviceTypeDto.WATCH
    Device.TYPE_PHONE -> DeviceTypeDto.PHONE
    Device.TYPE_SCALE -> DeviceTypeDto.SCALE
    Device.TYPE_RING -> DeviceTypeDto.RING
    Device.TYPE_FITNESS_BAND -> DeviceTypeDto.FITNESS_BAND
    Device.TYPE_CHEST_STRAP -> DeviceTypeDto.CHEST_STRAP
    Device.TYPE_HEAD_MOUNTED -> DeviceTypeDto.HEAD_MOUNTED
    Device.TYPE_SMART_DISPLAY -> DeviceTypeDto.SMART_DISPLAY
    else -> DeviceTypeDto.UNKNOWN
}

/**
 * Converts a [MetadataDto] to a Health Connect [Metadata] object.
 *
 * Note: [MetadataDto.lastModifiedTime] is managed by Health Connect and not set during writes.
 * The Metadata constructor is internal, so we use the appropriate factory method based on
 * the recording method. This will be populated by Health Connect when records are written.
 */
internal fun MetadataDto.toHealthConnect(): Metadata {
    val device = if (deviceType != null || deviceManufacturer != null || deviceModel != null) {
        Device(
            manufacturer = deviceManufacturer,
            model = deviceModel,
            type = deviceType?.toHealthConnect() ?: Device.TYPE_UNKNOWN,
        )
    } else {
        null
    }
    val clientRecordVersion = clientRecordVersion ?: 0L

    return when (recordingMethod) {
        RecordingMethodDto.ACTIVELY_RECORDED -> {
            requireNotNull(device) {
                "Device must be specified when using ACTIVELY_RECORDED recording method"
            }
            if (clientRecordId != null) {
                Metadata.activelyRecorded(device, clientRecordId, clientRecordVersion)
            } else {
                Metadata.activelyRecorded(device)
            }
        }

        RecordingMethodDto.AUTOMATICALLY_RECORDED -> {
            requireNotNull(device) {
                "Device must be specified when using AUTOMATICALLY_RECORDED recording method"
            }
            if (clientRecordId != null) {
                Metadata.autoRecorded(device, clientRecordId, clientRecordVersion)
            } else {
                Metadata.autoRecorded(device)
            }
        }

        RecordingMethodDto.MANUAL_ENTRY -> {
            if (clientRecordId != null) {
                Metadata.manualEntry(clientRecordId, clientRecordVersion, device)
            } else {
                Metadata.manualEntry(device)
            }
        }

        RecordingMethodDto.UNKNOWN -> {
            if (clientRecordId != null) {
                Metadata.unknownRecordingMethod(clientRecordId, clientRecordVersion, device)
            } else {
                Metadata.unknownRecordingMethod(device)
            }
        }
    }
}

/**
 * Converts a Health Connect [Metadata] object to a [MetadataDto].
 */
internal fun Metadata.toDto(): MetadataDto = MetadataDto(
    dataOrigin = dataOrigin.packageName,
    recordingMethod = recordingMethod.toRecordingMethodDto(),
    lastModifiedTime = lastModifiedTime.toEpochMilli(),
    clientRecordId = clientRecordId,
    clientRecordVersion = clientRecordVersion,
    deviceType = device?.type?.toDeviceTypeDto(),
    deviceManufacturer = device?.manufacturer,
    deviceModel = device?.model,
)
